dbms/src/Interpreters/Join.cpp

```diff
@@ -120,6 +120,34 @@ Join::Type Join::chooseMethod(const ColumnRawPtrs & key_columns, Sizes & key_siz
 }
 
 
+struct ASOFSplit
+{
+    ColumnRawPtrs key_columns;
+    Sizes key_sizes;
+
+    ColumnRawPtrs asof_columns;
+    Sizes asof_sizes;
+};
+
+template<ASTTableJoin::Strictness STRICTNESS>
+ASOFSplit split_asof_columns(const ColumnRawPtrs & key_columns, const Sizes & key_sizes)
```

Comment:
Some comments:

1. it seems to me that key_columns and key_sizes should be kept consistent, eg. when creating a KeyGetter it seems odd to pass in columns and sizes of different lengths, so extracting the asof column should also pop from key_sizes. 
2. I see that you pass key_columns as mutable reference into extractAsofColumn, from which I infer that it's intended to modify the key_columns by popping off the final column. This is fine, but would require removing the constness of the key_columns from the entire call chain - is that what you intended? From a clean code perspective you would not expect the function that joins the tables to actually modify the key columns (and also, what if somewhere down the line someone calls it twice? the second call would give incorrect results as it pops the final value off the vector of columns), it's merely input to it.

Reply:
> is that what you intended?
No, I mean to extract only asof column and make changes with key_columns in caller code.

Reply:
Ok, but that would entail copying the key_columns (and key_sizes) as they are passed in by const reference, after which you would have to modify them (pop off the last value for the normal key_getter and construct a vector of size 1 with the asof_column). The vector of size 1 is needed because the KeyGetter ctor (for both asof and the exact join column) expects a `const std::vector<const IColumn *> &`. I could add an additional ctor for the asof keygetter that just expects a const IColumn* as none of the other fields are used?

Reply:
It's not principal. I just wont to say it took me 10 minutes to understand what ASOFSplit is used for. And that is me who was modified the code of joins time to time :)

Reply:
I have removed it now - it is indeed a bit cleaner just using functions.